### PR TITLE
Add Mac OS X port

### PR DESCRIPTION
**Pull request by gangliao**

### CODE REVIEW

File: paddle/trainer/tests/test_CompareSparse.cpp [hunk 1]
```diff
@@ -146,12 +146,12 @@ TEST(compareSparse, remote_cpu) {
 TEST(compareSparse, cpu10_local_vs_remote) {
   FLAGS_local = 1;  // disable remote sparse update in parameter config
   std::vector<ParameterPtr> localParameters =
-      trainerOnePassTest(configFile1, true, 10);
+      trainerOnePassTest(configFile1, true, 2);
```
Comment (emailweixu) [inline paddle/trainer/tests/test_CompareSparse.cpp]: why can't "10" work?

File: paddle/api/paddle_ld_flags.py [hunk 2]
```diff
@@ -56,9 +69,9 @@ def parent_dir_str(self):
 
         def libs_str(self):
             libs = [
-                "-Wl,--whole-archive",
+                whole_start,
```
Comment (gangliao) [inline paddle/api/paddle_ld_flags.py]: All already done this via "-Wl, all_load" in line 48, /Users/liaogang/baidu/Paddle/paddle/setup.py.in.

File: paddle/gserver/gradientmachines/NeuralNetwork.cpp [hunk 3]
```diff
@@ -385,17 +385,10 @@ void NeuralNetwork::setOutputGrad(const std::vector<Argument>& args) {
   }
 }
 
-extern NeuralNetwork* newCustomNeuralNetwork(
-    const std::string& name, NeuralNetwork* network) __attribute__((weak));
-
 NeuralNetwork* NeuralNetwork::newNeuralNetwork(
     const std::string& name,
     NeuralNetwork* rootNetwork) {
-  if (newCustomNeuralNetwork) {
```
Comment (gangliao) [inline paddle/gserver/gradientmachines/NeuralNetwork.cpp]: use cmake to fix this   
```cmake
 if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang")
            list(APPEND LIBS "-undefined dynamic_lookup")
        endif()
```

File: doc/build/build_from_source.md [hunk 4]
```diff
@@ -139,3 +143,129 @@ And if you set WITH_SWIG_PY=ON, you have to install related python predict api a
...
+make -j `nproc` && make install
+# PaddlePaddle installation path
+export PATH=<path to install>/bin:$PATH
+```
+**Note**
+
+And if you set WITH_SWIG_PY=ON, you have to install related python predict api at the same time:
+
+```bash
+sudo pip install <path to install>/opt/paddle/share/wheels/*.whl
```
Comment (reyoung) [inline doc/build/build_from_source.md]: Or just run `sudo paddle version`

File: paddle/trainer/tests/test_CompareSparse.cpp (hunk 1 above)
Comment (gangliao) [inline paddle/trainer/tests/test_CompareSparse.cpp]: When PaddlePaddle linked with Openblas on MAC OS X,  run unit test triggers "35: BLAS : Program is Terminated. Because you tried to allocate too many memory regions. ", I guess the problem is from openblas.

File: doc/build/build_from_source.md [hunk 5]
```diff
@@ -139,3 +143,129 @@ And if you set WITH_SWIG_PY=ON, you have to install related python predict api a
...
+cmake -DWITH_GPU=ON -DWITH_DOC=OFF -DCMAKE_INSTALL_PREFIX=<path to install> ..
+# please use sudo make install, if you want
+# to install PaddlePaddle into the system
+make -j `nproc` && make install
+# PaddlePaddle installation path
+export PATH=<path to install>/bin:$PATH
+```
+**Note**
+
+And if you set WITH_SWIG_PY=ON, you have to install related python predict api at the same time:
```
Comment (reyoung) [inline doc/build/build_from_source.md]: Not only in WITH_SWIG_PY=ON. The `python` module is also need be installed.

Paddle will automatically install python dependencies at first time user run paddle commands, such as `paddle version`, `paddle train`. It may require `sudo` privileges.

File: paddle/api/Matrix.cpp [hunk 6]
```diff
@@ -95,7 +95,7 @@ float Matrix::get(size_t x, size_t y) const throw(RangeError) {
 }
 
 void Matrix::set(size_t x, size_t y, float val) throw(RangeError,
-                                                      UnsupportError) {
+                                                     UnsupportError) {
```
Comment (reyoung) [inline paddle/api/Matrix.cpp]: Should not change this line.

File: paddle/api/Matrix.cpp [hunk 7]
```diff
@@ -239,7 +239,7 @@ void Matrix::toNumpyMatInplace(float** view_data, int* dim1,
 }
 void Matrix::copyToNumpyMat(float** view_m_data, int* dim1,
                             int* dim2) throw(UnsupportError) {
-  static_assert(sizeof(paddle::real) == sizeof(float),
+  static_assert(sizeof(float) == sizeof(float),
```
Comment (gangliao) [inline paddle/api/Matrix.cpp]: fixed

File: paddle/api/Matrix.cpp [hunk 8]
```diff
@@ -251,12 +251,12 @@ void Matrix::copyToNumpyMat(float** view_m_data, int* dim1,
     if (auto cpuMat = dynamic_cast<paddle::CpuMatrix*>(m->mat.get())) {
       auto src = cpuMat->getData();
       auto dest = *view_m_data;
-      std::memcpy(dest, src, sizeof(paddle::real) * (*dim1) * (*dim2));
+      std::memcpy(dest, src, sizeof(float) * (*dim1) * (*dim2));
```
Comment (gangliao) [inline paddle/api/Matrix.cpp]: fixed

File: paddle/api/Matrix.cpp [hunk 9]
```diff
@@ -251,12 +251,12 @@ void Matrix::copyToNumpyMat(float** view_m_data, int* dim1,
...
     } else if (auto gpuMat = dynamic_cast<paddle::GpuMatrix*>(m->mat.get())) {
       auto src = gpuMat->getData();
       auto dest = *view_m_data;
       hl_memcpy_device2host(dest, src,
-                            sizeof(paddle::real) * (*dim1) * (*dim2));
+                            sizeof(float) * (*dim1) * (*dim2));
```
Comment (gangliao) [inline paddle/api/Matrix.cpp]: fixed

File: paddle/cuda/include/hl_device_functions.cuh [hunk 10]
```diff
@@ -16,28 +16,37 @@ limitations under the License. */
...
-        assumed = old;
-        old = atomicCAS(address_as_ull,
-                        assumed,
-                        __double_as_longlong(val +
-                        __longlong_as_double(assumed)));
-    } while (assumed != old);
-
-    return __longlong_as_double(old);
-}
+namespace paddle {
```
Comment (reyoung) [inline paddle/cuda/include/hl_device_functions.cuh]: This code is not related to osx port.

Could extract it to another PR?

File: paddle/math/tests/test_CpuGpuVector.cpp [hunk 11]
```diff
@@ -84,4 +84,10 @@ int main(int argc, char** argv) {
   return ret;
 }
 
+#else
```
Comment (reyoung) [inline paddle/math/tests/test_CpuGpuVector.cpp]: This lines may be can removed in mac osx.
Since the `whole-archive` was not correct before, this lines ware added.

I will give a PR to you, and check it is cool or not.

File: paddle/utils/Thread.h [hunk 12]
```diff
@@ -18,7 +18,18 @@ limitations under the License. */
...
+  pid_t tid = syscall(SYS_thread_selfid);
+#else
+  #ifndef __NR_gettid
+  #define __NR_gettid 224
+  #endif
+  pid_t tid = syscall(__NR_gettid);
+#endif
+  CHECK_NE(tid, -1);
+  return tid;
+}
```
Comment (gangliao) [inline paddle/utils/Thread.h]: fixed

File: paddle/utils/ThreadLocal.h [hunk 13]
```diff
@@ -156,7 +156,15 @@ class ThreadLocalD {
...
   void updateMap(T* p) {
-    pid_t tid = syscall(SYS_gettid);
+#if defined(__APPLE__) || defined(__OSX__)
+    pid_t tid = syscall(SYS_thread_selfid);
+#else
+    #ifndef __NR_gettid
+    #define __NR_gettid 224
+    #endif
+    pid_t tid = syscall(__NR_gettid);
+#endif
```
Comment (gangliao) [inline paddle/utils/ThreadLocal.h]: fixed

File: paddle/trainer/tests/test_Trainer.cpp [hunk 14]
```diff
@@ -90,7 +94,7 @@ TEST(checkGradient, multi) {
 TEST(checkGradient, hsigmoid) { checkGradientTest(configFile2, false, false); }
 
 TEST(checkGradient, chunk) {
-  EXPECT_EQ(0, system("python2 trainer/tests/gen_proto_data.py"));
+  EXPECT_EQ(0, system("python trainer/tests/gen_proto_data.py"));
```
Comment (reyoung) [inline paddle/trainer/tests/test_Trainer.cpp]: Is python2 not good?

But in some linux distribution, `python3` is default python

File: paddle/utils/PythonUtil.cpp [hunk 15]
```diff
@@ -144,12 +144,12 @@ PyObjectPtr createPythonClass(
     const std::map<std::string, std::string>& kwargs) {
   PyGuard guard;
   PyObjectPtr pyModule(PyImport_ImportModule(moduleName.c_str()));
-  LOG(INFO) << "createPythonClass moduleName.c_str:" << moduleName.c_str();
+  // LOG(INFO) << "createPythonClass moduleName.c_str:" << moduleName.c_str();
```
Comment (gangliao) [inline paddle/utils/PythonUtil.cpp]: fixed

File: paddle/utils/PythonUtil.h [hunk 16]
```diff
@@ -28,12 +34,14 @@ limitations under the License. */
...
-#warning "no _XOPEN_SOURCE defined in Python.h"
-#endif
+
+// #ifndef _POSIX_C_SOURCE
+// #warning "no _POSIX_C_SOURCE defined in Python.h"
+// #endif
+// #ifndef _XOPEN_SOURCE
+// #warning "no _XOPEN_SOURCE defined in Python.h"
+// #endif
+
```
Comment (gangliao) [inline paddle/utils/PythonUtil.h]: fixed

File: paddle/trainer/tests/test_Trainer.cpp (hunk 14 above)
Comment (gangliao) [inline paddle/trainer/tests/test_Trainer.cpp]: MAC OS X do not have python2. I will add some condition branch to avoid potential hazard.

File: cmake/cblas.cmake [hunk 17]
```diff
@@ -44,8 +44,8 @@ set(ATLAS_LIB_SEARCH_PATHS
         /usr/lib
         /usr/lib/blas/atlas
         /usr/lib/atlas
-        /usr/lib/atlas-base   # special for ubuntu 14.04.
-    )
```
Comment (gangliao) [inline cmake/cblas.cmake]: done

File: doc/build/build_from_source.md [hunk 18]
```diff
@@ -139,3 +143,129 @@ And if you set WITH_SWIG_PY=ON, you have to install related python predict api a
...
 pip install <path to install>/opt/paddle/share/wheels/*.whl
 ```
+## <span id="mac">Building on Mac OS X</span>
+
+### Prerequisites
+This guide is based on Mac OS X 10.11 (El Capitan). Note that if you are running an up to date version of OS X, 
+you will already have Python 2.7.10 and Numpy 1.8 installed.
+
+The best option is to use the package manager homebrew to handle installations and upgrades for you.
+To install homebrew, first open a terminal window (you can find Terminal in the Utilities folder in Applications), and issue the command:
```
Comment (gangliao) [inline doc/build/build_from_source.md]: fixed

File: doc/build/build_from_source.md [hunk 19]
```diff
@@ -139,3 +143,129 @@ And if you set WITH_SWIG_PY=ON, you have to install related python predict api a
...
+brew install glog gflags cmake protobuf openblas
+
+# Install google test on Mac OS X
+# Download gtest 1.7.0
+wget https://github.com/google/googletest/archive/release-1.7.0.tar.gz
+tar -xvf googletest-release-1.7.0.tar.gz && cd googletest-release-1.7.0
+# Build gtest
+mkdir build && cmake ..
+make
+# Install gtest library
```
Comment (gangliao) [inline doc/build/build_from_source.md]: After `make install`, print  'make install' is dangerous and not supported. Instead, see README for how to integrate Google Test into your build system.

File: doc/build/build_from_source.md [hunk 20]
```diff
@@ -139,3 +143,129 @@ And if you set WITH_SWIG_PY=ON, you have to install related python predict api a
...
+
+Finally, you can download source code and build:
+
+```bash
+git clone https://github.com/baidu/Paddle paddle
+cd paddle
+mkdir build
+cd build
+# you can add build option here, such as:    
+cmake -DWITH_GPU=ON -DWITH_DOC=OFF -DCMAKE_INSTALL_PREFIX=<path to install> ..
```
Comment (gangliao) [inline doc/build/build_from_source.md]: fixed

File: doc/build/build_from_source.md [hunk 21]
```diff
@@ -139,3 +143,129 @@ And if you set WITH_SWIG_PY=ON, you have to install related python predict api a
...
+CMake can find dependent libraries in system default paths firstly.
+After installing some optional libraries, corresponding build option will be on automatically (for instance, glog, gtest and gflags).
+If not found, you have to set following variables manually via CMake command (CUDNN_ROOT, ATLAS_ROOT, MKL_ROOT, OPENBLAS_ROOT).
+
+Here are some examples of CMake command with different options:
+
+**only cpu**
+
+```bash
+cmake -DWITH_GPU=OFF -DWITH_DOC=OFF
```
Comment (gangliao) [inline doc/build/build_from_source.md]: fixed

File: paddle/api/Matrix.cpp (hunk 6 above)
Comment (gangliao) [inline paddle/api/Matrix.cpp]: fixed

File: paddle/cuda/include/hl_device_functions.cuh (hunk 10 above)
Comment (gangliao) [inline paddle/cuda/include/hl_device_functions.cuh]: bug found in MAC OSX  when open WITH_DOUBLE

File: cmake/cblas.cmake (hunk 17 above)
Comment (gangliao) [inline cmake/cblas.cmake]: fixed

File: doc/build/build_from_source.md (hunk 18 above)
Comment (gangliao) [inline doc/build/build_from_source.md]: fixed

File: doc/build/build_from_source.md (hunk 19 above)
Comment (gangliao) [inline doc/build/build_from_source.md]: When PaddlePaddle linked with Openblas on MAC OS X, run unit test triggers "35: BLAS : Program is Terminated. Because you tried to allocate too many memory regions. ", I guess the problem is from openblas.

File: doc/build/build_from_source.md (hunk 21 above)
Comment (gangliao) [inline doc/build/build_from_source.md]: fixed

File: doc/build/build_from_source.md (hunk 20 above)
Comment (gangliao) [inline doc/build/build_from_source.md]: fixed

File: doc/build/build_from_source.md (hunk 4 above)
Comment (gangliao) [inline doc/build/build_from_source.md]: fixed

File: doc/build/build_from_source.md (hunk 5 above)
Comment (gangliao) [inline doc/build/build_from_source.md]: fixed

File: paddle/api/Matrix.cpp (hunk 6 above)
Comment (gangliao) [inline paddle/api/Matrix.cpp]: fixed

File: paddle/math/tests/test_CpuGpuVector.cpp (hunk 11 above)
Comment (gangliao) [inline paddle/math/tests/test_CpuGpuVector.cpp]: fixed
